subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/AbstractModuleDependency.java

```diff
@@ -48,6 +50,7 @@ public abstract class AbstractModuleDependency extends AbstractDependency implem
 
     private ImmutableAttributesFactory attributesFactory;
     private NotationParser<Object, Capability> capabilityNotationParser;
+    private ObjectFactory objectFactory;
```

Comment:
One of my main worries here is that we're adding a new field to every dependency in Gradle. How does this affect memory usage for very large builds? I tried running a perf test but I think we need to rebase this PR on master first. 

In any case, I think in practice most of these dependencies _already_ have a reference to an object factory, given that they are almost always instantiated with an ObjectFactory already (and thus have an implicit generated ObjectFactory field). 

If we determine that this does not affect memory-usage/performance we are probably fine but I think we should test this

Reply:
Perf tests passing so I guess this is fine. I think we can make improvements to the way we instantiate dependencies in general, though that's for another PR